### PR TITLE
Added getDistance function

### LINKED ISSUE #3: implement distance functions
numeric: euclidean
categorical: anti-dice (see knn_notes.md)

### PR DESCRIPTION
**Pull request by ethanbwaite**
- Added getDistance, weightedCategoricalDistance, weightedEuclideanDistance
- getDistance takes two case classes and calcualtes categorical and euclidean distance based on the objects' types, string's will be considered categorical
- Added example usage to the bottom of the file

Closes #3

### CODE REVIEW

File: source/src/main/scala/rides/App.scala [hunk 1]
```diff
@@ -34,9 +35,46 @@ case class Weather(
...
+      case (v1: Double, v2: Double) => pow(v1 - v2, 2).toDouble
+      case (v1: Float, v2: Float) => pow(v1 - v2, 2).toDouble
+      case _ => 0.0
+    }).sum
+    val numericFieldCount = values1.map({
+      case v: Int => 1
+      case v: Double => 1
+      case v: Float => 1
+      case _ => 0
+    }).reduce((x, y) => x+y)
```
Comment (lhmcgann) [inline source/src/main/scala/rides/App.scala]: do we need to use reduce or can we just use sum again?

File: source/src/main/scala/rides/App.scala [hunk 2]
```diff
@@ -34,9 +35,46 @@ case class Weather(
...
+
+  def weightedCategoricalDistance(values1: List[Any], values2: List[Any]): Double = {
+    val numberOfMismatches = (values1 zip values2).map({
+      case (v1: String, v2: String) => if (v1 != v2) 1 else 0
+      case _ => 0
+    }).sum
+    val categoricalFieldCount = values1.map({
+      case v: String => 1
+      case _ => 0
+    }).reduce((x, y) => x+y)
```
Comment (lhmcgann) [inline source/src/main/scala/rides/App.scala]: same: do we need to use reduce or can we just use sum again?

File: source/src/main/scala/rides/App.scala [hunk 3]
```diff
@@ -34,9 +35,46 @@ case class Weather(
...
   val wind: Float
-);
+)
 
 object App {
+  def weightedEuclideanDistance(values1: List[Any], values2: List[Any]): Double = {
+    val sumOfSquaredDifferences = (values1 zip values2).map({
+      case (v1: Int, v2: Int) => pow(v1 - v2, 2).toDouble
+      case (v1: Double, v2: Double) => pow(v1 - v2, 2).toDouble
+      case (v1: Float, v2: Float) => pow(v1 - v2, 2).toDouble
```
Comment (lhmcgann) [inline source/src/main/scala/rides/App.scala:47]: should we just pick float or double? why use both?

File: source/src/main/scala/rides/App.scala [hunk 4]
```diff
@@ -34,9 +35,69 @@ case class Weather(
...
+    val dist = numberOfMismatches / values1.length.toDouble
+
+    if (DEBUG) {
+      println(s"Mismatches: ${numberOfMismatches} / ${categoricalFieldCount} --> dist = " + f"${dist}%.2f")
+    }
+
+    return dist
+  }
+
+  def getDistance(row1: Product, row2: Product): Double = {
```
Comment (lhmcgann) [inline source/src/main/scala/rides/App.scala:73]: this is just a question, but what is a Product?

**Review by lhmcgann — COMMENTED** [review]
Looks good! I added just a few questions in the review comments and a few comments in the actual code (plus one small change to categorical calculation)

File: source/src/main/scala/rides/App.scala (hunk 1 above)
Comment (ethanbwaite) [reply source/src/main/scala/rides/App.scala]: sum is a better idea

File: source/src/main/scala/rides/App.scala (hunk 3 above)
Comment (ethanbwaite) [reply source/src/main/scala/rides/App.scala:47]: I saw Ethan had floats and doubles and I didn't think about it. I think just doubles are fine unless theres an outstanding reason for both

File: source/src/main/scala/rides/App.scala (hunk 4 above)
Comment (ethanbwaite) [reply source/src/main/scala/rides/App.scala:73]: It's the supertype for the `case` class. It contains the `productIterator` field which stores a list of all the attributes of the object. So `getDistance` can be generic and we can change the Ride and Weather object attributes without updating the distance functions

File: source/src/main/scala/rides/App.scala (hunk 3 above)
Comment (zimeg) [reply source/src/main/scala/rides/App.scala:47]: ah i was copying from [data_description.md](https://github.com/ethanbwaite/csc369/blob/main/input/data_description.md). should make no difference!

**Pull request merged**